packages/core/src/loaders/progressiveLoader.ts

```diff
@@ -0,0 +1,382 @@
+  nearbyItem;
+};
+
+export type ProgressiveRequest = {
+  imageId: string;
+  stage: RetrieveStage;
+  next?: ProgressiveRequest;
+  nearbyRequests?: NearbyRequest[];
+};
+
+/** Interleaves the values according to the stages definition */
+function interleave(
+  requests: string[],
+  retrieveConfiguration: IRetrieveConfiguration,
+  stageStatus: Map<string, StageStatus>
+) {
+  const { stages } = retrieveConfiguration;
+  const interleaved = new Array<ProgressiveRequest>();
+  // Maps image id to the LAST progressive request - to allow tail append
+  const imageRequests = new Map<string, ProgressiveRequest>();
+
+  const addStageInstance = (stage, position) => {
+    const index =
+      position < 0
+        ? requests.length + position
+        : position < 1
+        ? Math.floor((requests.length - 1) * position)
+        : position;
+    const imageId = requests[index];
+    if (!imageId) {
+      throw new Error(`No value found to add to requests at ${position}`);
+    }
+    const request: ProgressiveRequest = {
+      imageId,
+      stage,
+      nearbyRequests: findNearbyRequests(index, requests, stage),
+    };
+    addStageStatus(stageStatus, stage);
+    const existingRequest = imageRequests.get(imageId);
+    if (existingRequest) {
+      existingRequest.next = request;
+    } else {
+      interleaved.push(request);
+    }
+    imageRequests.set(imageId, request);
+  };
+
+  for (const stage of stages) {
+    const indices =
+      stage.positions ||
+      decimate(requests, stage.decimate || 1, stage.offset ?? 0);
+    indices.forEach((index) => addStageInstance(stage, index));
+  }
+  return interleaved;
+}
+
+/**
+ * Finds nearby requests to fulfill to show the merge information earlier.
+ * @param index - to use as the base value
+ * @param requests - set of image ids to request
+ * @param stage - to find information from
+ * @returns Array of nearby frames to fill when the main stage is done
+ */
+function findNearbyRequests(
+  index: number,
+  requests: string[],
+  stage
+): NearbyRequest[] {
+  const nearby = new Array<NearbyRequest>();
+  if (!stage.nearbyFrames) {
+    return nearby;
+  }
+  for (const nearbyItem of stage.nearbyFrames) {
+    const nearbyIndex = index + nearbyItem.offset;
+    if (nearbyIndex < 0 || nearbyIndex >= requests.length) {
+      continue;
+    }
+    nearby.push({
+      itemId: requests[nearbyIndex],
+      nearbyItem,
+      status: nearbyItem.status,
+    });
+    if (nearbyItem.linearOffset !== undefined) {
+      const linearIndex =
+        nearbyItem.linearOffset !== undefined &&
+        nearbyItem.linearOffset + nearbyItem.offset;
+      if (linearIndex >= 0 && linearIndex < requests.length) {
+        nearby[nearby.length - 1].linearId = requests[linearIndex];
+      }
+    }
+  }
+
+  return nearby;
+}
+
+/** Actually fills the nearby frames from the given frame */
+function fillNearbyFrames(
+  listener: ProgressiveListener,
+  imageStatus: Map<string, ImageStatus>,
```

Comment:
this is confusing again, imageStatus is a map (?) of string and imageStatus? so is this effectively imageStatusMap?